### ISSUE TITLE
bug generating history for git repos with symlink to itself

### ISSUE THREAD

**Issue by ilovemaui** [oracle/opengrok#1315]
I have noticed with the new Opengrok-0.13-rc5 a bug generating history for some git repos. It seems to be happening for git repos that have a symlink in them. here is a example error:

```
Dec 29, 2016 6:18:11 PM org.opensolaris.opengrok.history.HistoryGuru createCache
WARNING: An error occured while creating cache for mydir (GitRepository)
java.lang.StringIndexOutOfBoundsException: String index out of range: -1
        at java.lang.String.substring(String.java:1931)
        at org.opensolaris.opengrok.history.FileHistoryCache.isRenamedFile(FileH
istoryCache.java:141)
        at org.opensolaris.opengrok.history.FileHistoryCache.store(FileHistoryCa
che.java:394)
        at org.opensolaris.opengrok.history.Repository.createCache(Repository.ja
va:337)
        at org.opensolaris.opengrok.history.HistoryGuru.createCache(HistoryGuru.
java:509)
        at org.opensolaris.opengrok.history.HistoryGuru.access$000(HistoryGuru.j
ava:54)
        at org.opensolaris.opengrok.history.HistoryGuru$1.run(HistoryGuru.java:5
60)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:51
1)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.
java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor
.java:617)
        at java.lang.Thread.run(Thread.java:745)
```

**Comment by vladak** [comment]
Based on off-line conversation I reproduce the problem as follows:

```
mkdir gitcycle
cd gitcycle
git init
ln -s . foo
git add foo
git commit foo
```

**Comment by vladak** [comment]
Not sure why this code IndexDatabase.java#accept() did not help:

```
731            File f1 = parent.getCanonicalFile();
732            File f2 = file.getCanonicalFile();
733            if (f1.equals(f2)) {
734                LOGGER.log(Level.INFO, "Skipping links to itself...: {0} {1}",
735                        new Object[]{parent.getAbsolutePath(), file.getAbsolutePath()});
736                return false;
737            }
```

also, there is `IndexDatabase.java#acceptSymlink()`

**Comment by tulinkry** [comment]
As we discovered yesterday the history generation does not respect neither the ignore files/dirs nor the accepted symlinks. It never reaches this code in the history phase. The history generation reaches the point where it tries to store a history for a symlink to itself causing the substring operation to fail.

**Comment by tulinkry** [comment]
fixed with 8ed071e

**Comment by vladak** [comment]
The ignored files during repository handling were solved in #3137. The symlink handling also improved quite a bit. Hence closing.
